aodncore/pipeline/steps/harvest.py

```diff
@@ -149,75 +127,100 @@ def executor_conversion(executor):
-                matched_file_list = [mf.dest_path for mf in matched_files]
+                self.execute_talend(self._config.trigger_config[harvester]['exec'], matched_files, talend_base_dir)
 
-                self.execute_talend(self._config.trigger_config[harvester]['exec'], matched_file_list, talend_base_dir)
+            files_to_delete = matched_files.filter_by_bool_attribute('pending_store_deletion')
+            if files_to_delete:
+                self.upload_runner.run(files_to_delete)
 
-            for mf in matched_files:
-                mf.is_harvested = True
+    def run_undo_deletions(self, harvester_map):
+        """Function to un-harvest and undo stored files as appropriate in the case of errors
+        Operates in newly created temporary directory as talend requires a non-existant file to perform un-harvesting
 
-            files_to_delete = matched_files.filter_by_bool_attribute('pending_store_deletion')
+
+        :param harvester_map: mapping of harvesters to the PipelineFileCollection to operate on
+        """
+        for harvester, matched_files in harvester_map.items():
+            with TemporaryDirectory(prefix='talend_base', dir=self.tmp_base_dir) as talend_base_dir:
+                self.execute_talend(self._config.trigger_config[harvester]['exec'], matched_files, talend_base_dir,
+                                    'is_harvest_undone')
+
+            files_to_delete = matched_files.filter_by_bool_attributes_and('pending_undo', 'is_stored')
             if files_to_delete:
                 self.upload_runner.run(files_to_delete)
 
     def run_additions(self, harvester_map, tmp_base_dir):
+        """Function to harvest and upload files using the appropriate file upload runner
+        Operates in newly created temporary directory and creates symlink between source and destination file
+        Talend will then operate on the destination file (symlink)
+
+
+        :param harvester_map: mapping of harvesters to the PipelineFileCollection to operate on
+        :param tmp_base_dir: temporary directory base for talend operation
+        """
         for harvester, matched_files in harvester_map.items():
             with TemporaryDirectory(prefix='talend_base', dir=tmp_base_dir) as talend_base_dir:
                 for pf in matched_files:
                     self.create_symlink(talend_base_dir, pf.src_path, pf.dest_path)
 
-                matched_file_list = [mf.dest_path for mf in matched_files]
-
-                self.execute_talend(self._config.trigger_config[harvester]['exec'], matched_file_list, talend_base_dir)
-
-            for mf in matched_files:
-                mf.is_harvested = True
+                try:
+                    self.execute_talend(self._config.trigger_config[harvester]['exec'], matched_files, talend_base_dir)
+                except Exception:
+                    self.undo_processed_files()
```

Comment:
I think we need to add the current `{harvester: matched_files}` to the `harvested_file_map` before calling the undo, as there may have been files successfully harvested before the error happened!

In an ideal world one harvester run would be an atomic transaction, but in reality most harvesters commit to the db after each file :-1: :-1: